File: Java/Easy/Accepted/verifying-an-alien-dictionary_7.java
class Solution {
    public boolean isAlienSorted(String[] words, String order) {
        
        int[] charOrder = new int[26];
        for (int i = 0; i < order.length(); i++) {
            charOrder[order.charAt(i) - 'a'] = i;
        }
        for (int i = 0; i < words.length - 1; i++) {
            if (!isSorted(words[i], words[i + 1], charOrder)) {
                return false;
            }
        }
        return true;
    }
    
    private boolean isSorted(String s1, String s2, int[] charOrder) {
        for (int i = 0; i < s1.length(); i++) {
            if (i == s2.length()) {
                return false;
            }
            if (charOrder[s1.charAt(i) - 'a'] < charOrder[s2.charAt(i) - 'a']) {
                return true;
            }
            if (charOrder[s1.charAt(i) - 'a'] > charOrder[s2.charAt(i) - 'a']) {
                return false;
            }
        }
        return true;
    }
}

File: Rust/Easy/Accepted/verifying-an-alien-dictionary_7.rs
impl Solution {
    pub fn is_alien_sorted(words: Vec<String>, order: String) -> bool {
        
        let order = order.into_bytes();
        let mut map = vec![0; 26];
        for (i, c) in order.iter().enumerate() {
            map[*c as usize - 97] = i;
        }
        let mut words = words.iter().map(|w| {
            w.as_bytes().iter().map(|c| {
                map[*c as usize - 97]
            }).collect::<Vec<usize>>()
        }).collect::<Vec<Vec<usize>>>();
        for i in 1..words.len() {
            if words[i] < words[i - 1] {
                return false;
            }
        }
        true
    }
}